File: cpp/KalkCpp/Shape/Shape3D/RotationSolid/sphere.cpp
#include "sphere.h"
#include "math.h"
#include "../../Shape2D/circle.h"



Sphere::Sphere(const std::vector<Point3D>& vp3d, Circle* bs, ColorRGB* color):
    RotationSolid(vp3d, bs, "Sphere", color)
{

}

Sphere* Sphere::clone() const
{
    return new Sphere(*this);
}

std::vector<Point3D> Sphere::buildSphere(Circle* p) const
{
    std::vector<Point3D> points = std::vector<Point3D>();

    for(std::vector<Point2D>::const_iterator it = p->getPoints().begin(); it != p->getPoints().end(); ++it)
    {
        points.push_back(Point3D::p2dTop3d(*it));
    }

    return points;
}



Sphere::Sphere(Circle* bs, ColorRGB* color): RotationSolid(buildSphere(bs), bs, "Sphere", color)
{

}



Sphere* Sphere::operator-(const Shape3D& s) const
{
    Circle* newbaseShape = static_cast<Circle*>(baseShape)->operator -(*s.getBaseShape());
    Sphere* newSphere = new Sphere(buildSphere(newbaseShape),newbaseShape, getColor()->operator -(*s.getColor()));
    return newSphere;
}



Sphere* Sphere::operator+(const Shape3D& s) const
{
    Circle* newbaseShape = static_cast<Circle*>(baseShape)->operator +(*s.getBaseShape());
    Sphere* newSphere = new Sphere(buildSphere(newbaseShape),newbaseShape, getColor()->operator +(*s.getColor()));
    return newSphere;
}



Sphere* Sphere::operator*(const Shape3D& s) const
{
    Circle* newbaseShape = static_cast<Circle*>(baseShape)->operator *(*s.getBaseShape());
    Sphere* newSphere = new Sphere(buildSphere(newbaseShape),newbaseShape, getColor()->operator +(*s.getColor()));
    return newSphere;
}



Sphere* Sphere::operator/(const Shape3D& s) const
{
    Circle* newbaseShape = static_cast<Circle*>(baseShape)->operator /(*s.getBaseShape());
    Sphere* newSphere = new Sphere(buildSphere(newbaseShape),newbaseShape, getColor()->operator +(*s.getColor()));
    return newSphere;
}



double Sphere::diameter() const
{
    Circle* tmp = static_cast<Circle*>(baseShape);
    return tmp->diameter();
}


double Sphere::ray() const
{
    return static_cast<Circle*>(baseShape) ->ray();
}

double Sphere::volume() const
{
    return (4 * M_PI * pow(ray(),3)) / 3;
}



double Sphere::totalSurface() const
{
    return 4 * M_PI * pow(ray(),2);
}

double Sphere::circumference() const
{
    Circle* tmp = static_cast<Circle*>(baseShape);
    return tmp->circumference();
}

File: Java/src/Shape/Shape3D/RotationSolid/Sphere.java
package Shape.Shape3D.RotationSolid;

import Color.ColorRGB;
import Shape.Point2D;
import Shape.Point3D;
import Shape.Shape2D.Circle;
import Shape.Shape3D.Shape3D;

import java.util.ArrayList;

public class Sphere extends RotationSolid {

    public Sphere(ArrayList<Point3D> p, Circle bShape) {
        super(p, bShape, "Sphere");
    }

    public Sphere(ArrayList<Point3D> p, Circle bShape, ColorRGB color) {
        super(p, bShape, "Sphere", color);
    }

    public Sphere(Circle c) {
        super(null, c, "Sphere");
        points = buildSphere(c);
    }

    public Sphere(Circle c, ColorRGB color) {
        super(null, c, "Sphere", color);
        points = buildSphere(c);
    }

    public Sphere(Sphere s) {
        super(s);
    }

    private ArrayList<Point3D> buildSphere(Circle c) {
        ArrayList<Point3D> points = new ArrayList<>();
        for (Point2D p2d : c.getPoints()) {
            points.add(Point3D.p2dTop3d(p2d));
        }

        return points;
    }

    public Sphere sum(Shape3D other) {
        Circle newbaseShape = (Circle) baseShape.sum(other.getBaseShape());
        return new Sphere(
                buildSphere(newbaseShape), newbaseShape, getColor().sum(other.getColor()));
    }

    public Sphere sub(Shape3D other) {
        Circle newbaseShape = (Circle) baseShape.sub(other.getBaseShape());
        return new Sphere(
                buildSphere(newbaseShape), newbaseShape, getColor().sub(other.getColor()));
    }

    public Sphere mul(Shape3D other) {
        Circle newbaseShape = (Circle) baseShape.mul(other.getBaseShape());
        return new Sphere(
                buildSphere(newbaseShape), newbaseShape, getColor().mul(other.getColor()));
    }

    public Sphere div(Shape3D other) {
        Circle newbaseShape = (Circle) baseShape.div(other.getBaseShape());
        return new Sphere(
                buildSphere(newbaseShape), newbaseShape, getColor().sum(other.getColor()));
    }

    public double diameter() {
        Circle c = (Circle) baseShape;
        return c.diameter();
    }

    public double ray() {
        Circle c = (Circle) baseShape;
        return c.ray();
    }

    public double circumference() {
        Circle c = (Circle) baseShape;
        return c.circumference();
    }

    public double volume() {
        return (4 * Math.PI * Math.pow(ray(), 2)) / 3;
    }

    public double totalSurface() {
        return 4 * Math.PI * Math.pow(ray(), 2);
    }

    public String toString() {
        return super.toString() + "Ray " + ray() + "\nVolume " + volume() +
                "\nDiameter " + diameter() + "\nTotalSurface " + totalSurface();
    }
}
